### ISSUE TITLE
bug: `home.file` confusingly breaks nix patterns about strings and paths

### ISSUE THREAD

**Issue by tejing1** [nix-community/home-manager#3032]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Issue description

There is a near-universal pattern in the nix language that strings containing file paths are runtime-resolved while nix path types are buildtime-resolved. This is naturally adhered to almost everywhere due to the way antiquoting handles the 2 cases.

For example, in nixos, the options
```nix
{
  environment.etc."examplefile1".source = /path/to/examplefile;
  environment.etc."examplefile2".source = "/path/to/examplefile";
}
```
will result in these link targets
```
$ readlink -f /etc/examplefile*
/nix/store/1rj4006ms9zxrvygf44pzqzmbflqihln-examplefile
/path/to/examplefile
```

Similar behavior exists for most options which take an arbitrary file path, such as the home-manager option `services.mpd.musicDirectory`.

However, with home-manager's `home.file` option,
```nix
{
  home.file."examplefile1".source = /path/to/examplefile;
  home.file."examplefile2".source = "/path/to/examplefile";
}
```
will result in these link targets
```
$ readlink -f ~/examplefile*
/nix/store/1rj4006ms9zxrvygf44pzqzmbflqihln-examplefile
/nix/store/1rj4006ms9zxrvygf44pzqzmbflqihln-examplefile
```
which is very surprising/confusing, at least to me and others I know who use home-manager. I'm aware that the `mkOutOfStoreSymlink` function can be used to get the other behavior if desired, but breaking with the general pattern in the nix language and requiring a home-manager-specific workaround to get the same functionality back doesn't seem like it has an upside.

The code responsible for this behavior is https://github.com/nix-community/home-manager/blob/7c04351a57bb36133e6cf3b609e573b9e18b8eec/modules/files.nix#L17-L24

Looking at the commit message for this commit, I understand that at least part of the reason the code exists is to work around NixOS/nix#6515, however that could be done without deviating from the normal expectations about the behavior of strings and paths.

### Maintainer CC

@rycee

### System information

```markdown
- system: `"x86_64-linux"`
 - host os: `Linux 5.15.43, NixOS, 22.05 (Quokka), 22.05.20220530.a634c8f`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.8.1`
 - nixpkgs: `/etc/nix/path/nixpkgs`
```

**Comment by lilyinstarlight** [comment]
I will also chime in to concur that this is unintuitive and surprising behavior, given most of the rest of the Nix ecosystem can handle modules with options like `[option]."[file]".source = "arbitrary string";` without attempting to copy anything to the store

(Also I may have been the "some people" @NobbZ described above that created an activation script to link them, since I never knew about `mkOutOfStoreSymlink` 😅)

**Comment by rycee** [comment]
Fair enough that it is confusing. I'm somewhat reluctant to silently accept strings as out of store items, it would be nicer to error out in that case or show a warning to use `mkOutOfStoreSymlink` instead. I'll have to think about how to achieve that.

@tejing1 Do you have any concrete example of "… however that could be done without deviating from the normal expectations about the behavior of strings and paths" that I could have a look at?

**Comment by tejing1** [comment]
> I'm somewhat reluctant to silently accept strings as out of store items, though

Can you explain your reasoning a little more here? The cost of "but we're different" is rather high, so it seems like it should require a very clear reason to defy what the rest of the ecosystem does.

I'd love to be able to tell nixos/home-manager newbies "paths are always resolved at eval time, and strings at run time", but right now I always have to qualify that with "except for home-manager's `home.file` and `xdg.configFile` options, and no, I don't know why."

> Do you have any concrete example of "… however that could be done without deviating from the normal expectations about the behavior of strings and paths" that I could have a look at?

Something like
```nix
 sourceStorePath = file: 
   if builtins.isString file.source
   then file.source
   else let 
     sourcePath = toString file.source; 
     sourceName = config.lib.strings.storeFileName (baseNameOf sourcePath); 
   in 
     builtins.path { path = file.source; name = sourceName; }; 
```
should do it.

**Comment by rycee** [comment]
I think you still need to check for sources with context, something like

```nix
sourceStorePath = file:
  if builtins.isString file.source
  then file.source
  else let
    sourcePath = toString file.source;
    sourceName = config.lib.strings.storeFileName (baseNameOf sourcePath);
  in
    if builtins.hasContext sourcePath
    then file.source
    else builtins.path { path = file.source; name = sourceName; };
```

That said, this wouldn't distinguish between a literal string and something like `builtins.toFile "foo" "bar"`.

But

``` nix
if builtins.isString file.source && !(hasPrefix (builtins.storeDir))
```

maybe would do it. Then in the `if` branch one could throw or something an error that says to use `mkOutOfStoreSymlink`…

**Comment by rycee** [comment]
Tried out

```nix
sourceStorePath = file:
  if builtins.isString file.source && !(hasPrefix (builtins.storeDir) file.source)
  then throw ''
    Cannot specify a source path outside the Nix store, in this case

      "${file.source}"

    Please use a path literal instead.

    If you are really certain you want to have a source outside the Nix
    store then use

      config.lib.file.mkOutOfStoreSymlink "${file.source}"

    but please note that this will make your configuration non-reproducible.''
  else let
    sourcePath = toString file.source;
    sourceName = config.lib.strings.storeFileName (baseNameOf sourcePath);
  in
    if builtins.isString file.source || builtins.hasContext sourcePath
    then file.source
    else builtins.path { path = file.source; name = sourceName; };
```

and `home.file.blaaah.source = "/home/rycee/blah"` results in

```
error: Cannot specify a source path outside the Nix store, in this case

         "/home/rycee/blah"

       Please use a path literal instead.

       If you are really certain you want to have a source outside the Nix
       store then use

         config.lib.file.mkOutOfStoreSymlink "/home/rycee/blah"

       but please note that this will make your configuration non-reproducible.
```

which seems reasonably, I think. Should be possible to do this using `assertions` instead of `throw` as well so it wouldn't just abort the evaluation.

**Comment by tejing1** [comment]
> That said, this wouldn't distinguish between a literal string and something like builtins.toFile "foo" "bar".

It doesn't need to, since it just passes the value through if it's a string.

However I did realize upon thinking more about this that my version wouldn't handle a derivation like a `pkgs.writeScript` the same as yours does. Also, the name `sourceStorePath` is somewhat inaccurate. My updated suggestion:
```nix
 sourceString = file: 
   if builtins.isPath file.source
   then
     builtins.path {
       path = file.source;
       name = config.lib.strings.storeFileName (baseNameOf (toString file.source));
     }
   else "${file.source}";
```

> Linking directly to a file, in this case \${file.source}, causes your configuration to not be reproducible. If you want to make it reproducible then use a path literal. If you are certain you want to link directly to your file outside the Nix store then use 'config.lib.file.mkOutOfStoreSymlink "${file.source}"'.

So *that*'s the reason?... I would never have guessed that one. Ok, I understand the desire to warn people about this, though I personally think it's not necessary. However, please don't make us use `mkOutOfStoreSymlink` to get around it. How about having a home-manager option we can set to just turn off this warning behavior altogether and let it work like we expect for those of us who know how it works and actually want to do this?

It *is* quite normal to want to do this. Particularly if you run an ephemeral home. You need to create symlinks to your permanently stored state all the time, and `mkOutOfStoreSymlink` is quite clunky. Saying it "causes your configuration to not be reproducible" is a bit misleading given that kind of usecase. Perhaps something more like "This causes the target of the link not to be incorporated into your config and its generations and rollbacks, but instead remain stateful"?

**Comment by ncfavier** [comment]
You can drop ` || builtins.hasContext sourcePath`: only strings can have context.

**Comment by archer-65** [comment]
Not stale IMO
